File: csharp/1371.find-the-longest-substring-containing-vowels-in-even-counts.cs
/*
 * @lc app=leetcode id=1371 lang=csharp
 *
 * [1371] Find the Longest Substring Containing Vowels in Even Counts
 */

using System;
using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static IDictionary<char, int> Map = new Dictionary<char, int>() {
        { 'a', 0 },
        { 'e', 1 },
        { 'i', 2 },
        { 'o', 3 },
        { 'u', 4 }
    };

    public int FindTheLongestSubstring(string s) {
        int acc = 0, ret = 0, n = s.Length;
        var first = new Dictionary<int, int>();
        first[0] = -1;
        for (int i = 0; i < n; ++i) {
            if (Map.TryGetValue(s[i], out int j)) {
                acc ^= (1 << j);
            }
            if (first.TryGetValue(acc, out int p)) {
                ret = Math.Max(ret, i - p);
            } else {
                first[acc] = i;
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/1371.find-the-longest-substring-containing-vowels-in-even-counts.java
/*
 * @lc app=leetcode id=1371 lang=java
 *
 * [1371] Find the Longest Substring Containing Vowels in Even Counts
 */

// @lc code=start
class Solution {
    public int findTheLongestSubstring(String s) {
        int n = s.length(), ret = 0, acc = 0;
        int[] first = new int[1 << 5];
        first[0] = 1;
        for (int k = 0; k < n; ++k) {
            char c = s.charAt(k);
            if (c == 'a') {
                acc ^= (1 << 0);
            } else if (c == 'e') {
                acc ^= (1 << 1);
            } else if (c == 'i') {
                acc ^= (1 << 2);
            } else if (c == 'o') {
                acc ^= (1 << 3);
            } else if (c == 'u') {
                acc ^= (1 << 4);
            }
            if (first[acc] != 0) {
                ret = Math.max(ret, k + 2 - first[acc]);
            } else {
                first[acc] = k + 2;
            }
        }
        return ret;
    }
}
// @lc code=end

